crates/turbopack-dev/src/ecmascript/runtime.rs

```diff
@@ -0,0 +1,236 @@
+            chunk_group,
+        )
+        .into()])
+    }
+
+    #[turbo_tasks::function]
+    fn content(&self, origin_chunk: EcmascriptChunkVc) -> EcmascriptChunkRuntimeContentVc {
+        EcmascriptDevChunkContentVc::new(
+            origin_chunk,
+            self.chunking_context,
+            self.chunk_group,
+            Value::new(self.mode),
+        )
+        .into()
+    }
+
+    #[turbo_tasks::function]
+    async fn merge(
+        &self,
+        runtimes: Vec<EcmascriptChunkRuntimeVc>,
+    ) -> Result<EcmascriptChunkRuntimeVc> {
+        let Self {
+            chunking_context,
+            chunk_group,
+            mut mode,
+        } = self;
+
+        let chunking_context = chunking_context.resolve().await?;
+        let chunk_group = if let Some(chunk_group) = chunk_group {
+            Some(chunk_group.resolve().await?)
+        } else {
+            None
+        };
+
+        for runtime in runtimes {
+            let Some(runtime) = EcmascriptDevChunkRuntimeVc::resolve_from(runtime).await? else {
+                bail!("cannot merge EcmascriptDevChunkRuntime with non-EcmascriptDevChunkRuntime");
+            };
+
+            let Self {
+                chunking_context: other_chunking_context,
+                chunk_group: other_chunk_group,
+                mode: other_mode,
+            } = &*runtime.await?;
+
+            let other_chunking_context = other_chunking_context.resolve().await?;
+            let other_chunk_group = if let Some(other_chunk_group) = other_chunk_group {
+                Some(other_chunk_group.resolve().await?)
+            } else {
+                None
+            };
+
+            if chunking_context != other_chunking_context {
+                bail!("cannot merge EcmascriptDevChunkRuntime with different chunking contexts",);
+            }
+
+            if chunk_group != other_chunk_group {
+                bail!("cannot merge EcmascriptDevChunkRuntime with different chunk groups",);
+            }
+
+            mode |= *other_mode;
+        }
+
+        Ok(EcmascriptDevChunkRuntime {
+            chunking_context,
+            chunk_group,
+            mode,
+        }
+        .cell()
+        .into())
+    }
+
+    #[turbo_tasks::function]
+    fn evaluated(&self) -> EcmascriptChunkRuntimeVc {
+        EcmascriptDevChunkRuntime {
+            chunking_context: self.chunking_context,
+            chunk_group: self.chunk_group,
+            mode: EcmascriptDevChunkRuntimeMode::RegisterAndEvaluate,
+        }
+        .cell()
+        .into()
+    }
+}
+
+#[turbo_tasks::value(serialization = "auto_for_input")]
+#[derive(Debug, Clone, Copy, PartialOrd, Ord, Hash)]
+pub(crate) enum EcmascriptDevChunkRuntimeMode {
+    /// The main runtime code will be included in the chunk and the main entries
+    /// will be evaluated as soon as the chunk executes.
+    RegisterAndEvaluate,
+    /// The chunk's entries will be registered within an existing runtime.
+    Register,
```

Comment:
This is a bit weird...

Assuming chunk A (main_entries: X) with RegisterAndEvaluate and chunk B (main_entries: Y) with Register will be merged. This would lead to X and Y being executed, while the unmerged versions would only execute X.

I think the entries for execution need to be stored in a separate list.

Reply:
Or give chunks a list of entries to execute and determine the runtime mode from the list.